### PR TITLE
[clang] Fix CXXNewExpr end source location for 'new struct S'

### LINKED ISSUE #35300: CXXNewExpr returns invalid end loc for valid code "new struct Foo;"
|  |  |
| --- | --- |
| Bugzilla Link | [35952](https://llvm.org/bz35952) |
| Version | trunk |
| OS | All |
| CC | @DougGregor |

## Extended Description 
Reproduce code:

```
void NewExprCheck::registerMatchers(MatchFinder *Finder) {
  Finder->addMatcher(cxxNewExpr().bind("x"), this);
}

void NewExprCheck::check(const MatchFinder::MatchResult &Result) {
  const auto *E = Result.Nodes.getNodeAs<CXXNewExpr>("x");
  E->getEndLoc().dump(Result.Context->getSourceManager());
}
```

For the following test code:

```
struct Foo {};

void f() {
  new struct Foo; // invalid end loc
  new Foo; // valid end loc
}
```

### PR DESCRIPTION
**Pull request by necto**
Fixes #35300 
Currently, `new struct S` fails to set any valid end source location because the token corresponding to `S` is consumed in `ParseClassSpecifier` and is not accessible in the `ParseDeclarationSpecifiers` that normally sets the end source location.

### CODE REVIEW

**Comment by necto** [comment]
The invalid end location affects the CSA diagnostics, as you can [see on CE](https://compiler-explorer.com/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGe1wAyeAyYAHI%2BAEaYxCDSAA6oCoRODB7evnoJSY4CQSHhLFExXLaY9jkMQgRMxARpPn4ldpgOKVU1BHlhkdGxttW19RlNA53B3YW9XACUtqhexMjsHAoExF4OANRCmyYA7FYaAIL7ACImAMyHRwBuqHjom0RxENO7B5shAO6bq%2BtbQkuFne52OHFmtE4AFZeH5uLxUJw3NZrL95otMLszBceKQCJpwbMANYgC4ADgAdGTqTTaTSAJz6TiSWEE0iIji8BQgDR4gmzOCwGCIEBMVbkShoFhxOjRUKsZaqMkANgAtMrJJtkAYjJsIGsvAwidNeJh8ERiA89PxBCIxOwpDJBIoVOoOFpSLommUWhUXAx3J4Gv4A10CkVMolkgIho1SFlowwwz1iqVym1RrG9M1WgJ2rVk5NU6sOlmRh1CxGZnjiJhMFaNBDoaz3QjOEdVpsAGp4TBfaKbJVqjVanXAPUGo1vCC4QgkLE40ibDzS2XEBczXj41vTWYITBMLAxV5Mjgs0gsWJmClQqQXSRQi44sySZV7STSOEejlcnl8nekIKIpSjK9BkBQEAgWuKBjgA%2BpORJ8HQBDRNyEARGyETBDUACenC4lhzDEDhADyETaL6%2BG8FKbCCCRDC0HhrakFgEReMAbhiLQ3LwixmAsIYwDiMx%2BC1q0NyYDxHqYKoLReChVHkIIZRsrQeARMQuEeFgbJrHgl68RJxARIkmCnPxglqUY/J8AYwAKD2fYkXEjCKTawiiOIjruS6ahsl6%2BiCSgKKWPo6ncpAsyoHEFQ8aqqqGGIOEAF6YqqAAaiWMal64Jcw2XRKq8wEHE8mXKcKGqAQCJGZaWARSeOZ%2BhArhlqQgTjOGUzxlGFRtQmFSVt1TUZqWQbDGmvqjQWnUptmmbjXGJYzfkc3Vgo6JLBITYcDCpBfm2HCDiq6qatqgkTusU56rOFobia/5aLupAkmY15vR9n1fVCp7npeL4UlwULYmYXBPlCkhcPS9IXPtbI/rYf7bk9gHChASDFaVBASpBqCrmB8psJwQ6naOF1cHsFIaIDprmiQVolO5dpedIPlKH5zEBV8mlxFRO17Qd7KcCR8lY5sqBUMdw5nWOl2GsaeorqBA7mDiD3I4Sv28JeT4Uk%2B%2BsG4bcPMQj3K8hrz0kpI14XMqt4XBokgaBcexA8qZinhcLbfpwW42UB6MgEQng4zU9m%2B6QYcKMohhlEICCoF8cK4lBBMKsTJ0jsAyDIJsYMUmYtNznV1qyMzDqs7Ivluh6AUsAITBoLdJAABJihYdYMEc8moM5jBmop9fME3dPEH3yfaw3aBVPgE/9LPEerLPMchLQ8eJ3Pwe0AAkugIBCA3ZDc0wvPwvz3uHQfRFS6T526vqV0KzOo8bkuStrvdfs7sSEiUu%2BUL0jJBoPYwCLj0ghvSZUWtjY%2B05Ijc2/IdqFxgYdL%2BT1ZhGSSM4SQQA%3D%3D)

File: clang/test/Misc/ast-source-ranges.cpp [hunk 1]
```diff
@@ -0,0 +1,5 @@
+// RUN: %clang_cc1 -ast-dump %s  2>&1 | FileCheck %s
```
Comment (hokein) [inline clang/test/Misc/ast-source-ranges.cpp:1]: For ast-dump lit tests, we usually place them in `clang/test/AST/` dir with name `ast-dump*.cpp` (or we can add it to the existing `ast-dump-expr.cpp`).

**Review by hokein — APPROVED** [review]
thanks, this looks good to me, a small nit.

**Comment by necto** [comment]
> `clang-tidy/checkers/modernize/make-unique.cpp` is failing for whatever reason. Otherwise LGTM.

Indeed, turns out this PR fixes #35300

File: clang/test/AST/ast-dump-expr.cpp [hunk 2]
```diff
@@ -583,3 +583,8 @@ void NonADLCall3() {
   f(x);
 }
 } // namespace test_adl_call_three
+
```
Comment (hokein) [inline clang/test/AST/ast-dump-expr.cpp:586]: nit: wrap the code within `namespace GH35300 { ...}`.

**Review by hokein — APPROVED** [review]
thanks, still good.

File: clang/test/AST/ast-dump-expr.cpp (hunk 2 above)
Comment (necto) [reply clang/test/AST/ast-dump-expr.cpp:586]: Done

**Pull request merged**